include/range/v3/view/drop_last.hpp

```diff
@@ -0,0 +1,159 @@
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/ericniebler/range-v3
+//
+
+#ifndef RANGES_V3_VIEW_DROP_LAST_HPP
+#define RANGES_V3_VIEW_DROP_LAST_HPP
+
+#include <type_traits>
+#include <meta/meta.hpp>
+#include <range/v3/range_concepts.hpp>
+#include <range/v3/view/all.hpp>
+#include <range/v3/view/view.hpp>
+#include <range/v3/view_interface.hpp>
+#include <range/v3/view_adaptor.hpp>
+#include <range/v3/utility/functional.hpp>
+
+namespace ranges
+{
+    inline namespace v3
+    {
+        template<typename Rng>
+        struct drop_last_bidirectional_view
+          : view_interface<drop_last_bidirectional_view<Rng>, range_cardinality<Rng>::value>
+        {
+            CONCEPT_ASSERT(ranges::BidirectionalRange<Rng>());
+        private:
+            friend range_access;
+            Rng rng;
+            std::size_t n;
+            detail::non_propagating_cache<sentinel_t<Rng>> sentinel;
+            void update_sentinel()
+            {
+                if (!sentinel)
+                {
+                    sentinel = ranges::prev(ranges::end(rng), n, ranges::begin(rng));
+                }
+            }
+        public:
+            drop_last_bidirectional_view() = default;
+            drop_last_bidirectional_view(Rng&& rng, std::size_t n)
+              : rng(std::move(rng)), n(n)
+            {}
+
+            iterator_t<Rng> begin()
+            {
+                return ranges::begin(rng);
+            }
+            sentinel_t<Rng> end()
+            {
+                update_sentinel();
+                return *sentinel;
+            }
+            Rng & base()
+            {
+                return rng;
+            }
+            Rng const & base() const
+            {
+                return rng;
+            }
+        };
+
+        template<typename Rng>
+        struct drop_last_forward_view
+          : view_adaptor<drop_last_forward_view<Rng>, Rng>
+        {
+            CONCEPT_ASSERT(ranges::ForwardRange<Rng>());
+        private:
+            friend range_access;
+            std::size_t n;
+
+            struct sentinel_adaptor;
+            class adaptor : public adaptor_base
+            {
+                friend sentinel_adaptor;
+                iterator_t<Rng> probe;
+            public:
+                adaptor() = default;
+                adaptor(const Rng& rng, std::size_t n)
+                {
+                    probe = ranges::next(ranges::begin(rng), n, ranges::end(rng));
+                }
+
+                void next(iterator_t<Rng>& it)
+                {
+                    ++it;
+                    ++probe;
+                }
+            };
+
+            struct sentinel_adaptor :  adaptor_base
+            {
+                bool empty(const iterator_t<Rng>&, const adaptor& ia, const sentinel_t<Rng>& s) const
```

Comment:
East `const`, please.

Reply:
What about enforcing format with clang-format, as final CI step? Something like this https://github.com/Sarcasm/run-clang-format

As a bonus, many IDE's understand .clang-format, and will highlight/reformat problems.

Reply:
Unfortunately, that's not possible. See https://github.com/ericniebler/range-v3/issues/802#issuecomment-450400841.
Clang-Format does provide ways to mark range-for emulating macros (`ForEachMacros`), block macros (`MacroBlockBegin`, `MacroBlockEnd`) and statement macros (`StatementMacros`), but they're not enough for the concept emulation macros.

Reply:
@johelegp can't it be just opt-outed for macroses entirely? 

In any case, I get it - it's already in TODO list.

Reply:
There's no option for doing that. As the linked comment says, you'd have to splash these `// ...`-disabling comments everywhere.